lib/asciidoctor/substitutors.rb

```diff
@@ -1345,158 +1338,89 @@ def resolve_pass_subs subs
     end
 
-    linenums_mode = nil
-    highlight_lines = nil
-
-    case highlighter
-    when 'coderay'
-      if (linenums_mode = (attr? 'linenums', nil, false) ? (@document.attributes['coderay-linenums-mode'] || :table).to_sym : nil)
-        start = 1 if (start = (attr 'start', nil, 1).to_i) < 1
-        if attr? 'highlight', nil, false
-          highlight_lines = resolve_lines_to_highlight source, (attr 'highlight', nil, false)
-        end
-      end
-      result = ::CodeRay::Duo[attr('language', :text, false).to_sym, :html, {
-        css: (@document.attributes['coderay-css'] || :class).to_sym,
-        line_numbers: linenums_mode,
-        line_number_start: start,
-        line_number_anchors: false,
-        highlight_lines: highlight_lines,
-        bold_every: false,
-      }].highlight source
-    when 'pygments'
-      lexer = ::Pygments::Lexer.find_by_alias(attr 'language', 'text', false) || ::Pygments::Lexer.find_by_mimetype('text/plain')
-      opts = { cssclass: 'pyhl', classprefix: 'tok-', nobackground: true, stripnl: false }
-      opts[:startinline] = !(option? 'mixed') if lexer.name == 'PHP'
-      unless (@document.attributes['pygments-css'] || 'class') == 'class'
-        opts[:noclasses] = true
-        opts[:style] = (@document.attributes['pygments-style'] || Stylesheets::DEFAULT_PYGMENTS_STYLE)
-      end
-      if attr? 'highlight', nil, false
-        unless (highlight_lines = resolve_lines_to_highlight source, (attr 'highlight', nil, false)).empty?
-          opts[:hl_lines] = highlight_lines.join ' '
-        end
-      end
-      # NOTE highlight can return nil if something goes wrong; fallback to source if this happens
-      # TODO we could add the line numbers in ourselves instead of having to strip out the junk
-      if (attr? 'linenums', nil, false) && (opts[:linenostart] = (start = attr 'start', 1, false).to_i < 1 ? 1 : start) &&
-          (opts[:linenos] = @document.attributes['pygments-linenums-mode'] || 'table') == 'table'
-        linenums_mode = :table
-        if (result = lexer.highlight source, options: opts)
-          result = (result.sub PygmentsWrapperDivRx, '\1').gsub PygmentsWrapperPreRx, '\1'
-        else
-          result = sub_specialchars source
-        end
-      elsif (result = lexer.highlight source, options: opts)
-        if PygmentsWrapperPreRx =~ result
-          result = $1
-        end
-      else
-        result = sub_specialchars source
-      end
+    doc_attrs = @document.attributes
+    syntax_hl_name = syntax_hl.name
+    if (linenums_mode = (attr? 'linenums', nil, false) ? (doc_attrs[%(#{syntax_hl_name}-linenums-mode)] || :table).to_sym : nil)
+      start_line_number = 1 if (start_line_number = (attr 'start', nil, 1).to_i) < 1
     end
+    highlight_lines = resolve_lines_to_highlight source, (attr 'highlight') if attr? 'highlight', nil, false
+
+    highlighted, source_offset = syntax_hl.highlight self, source, (attr 'language', 'text', false),
```

Comment:
`highlight` method is optional ?

Reply:
Correct. Asciidoctor doesn't call `highlight` for client-side syntax highlighters. Instead, it prepares the assets (docinfo) and the real work happens when the page loads. `highlight` is only used when the highlighting is part of the Asciidoctor conversion. The `highlight?` method tells Asciidoctor whether it will be using this method or not.

What's missing on this class are some API docs to explain this.

Reply:
Right I was missing that part `@subs[idx] = :highlight` and then `highlight_source` is only called `@subs` contains `:highlight`.
All good :wink:

Reply:
Now it should be even clearer since highlight? is determined whether to add the :highlight sub and whether the underlying library is available.